File: week-2/problem-sets/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    string text = get_string("Text: ");
    float L = ((float) count_letters(text) / count_words(text)) * 100.00;
    float S = ((float) count_sentences(text) / count_words(text)) * 100.00;
    int index = roundf((0.0588 * L) - (0.296 * S) - 15.8);
    if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

int count_letters(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            count += 1;
        }
    }
    return count;
}

int count_words(string text)
{
    int count =
        1; // Accounts for the last word (i.e. without count=1, 'how are you' counts as two whereas 'how are you ' counts as three.)
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == ' ' || text[i] == '\n')
        {
            while (text[i + 1] == ' ' || text[i + 1] == '\n') // Skips consecutive spaces
            {
                i++;
            }
            count += 1;
        }
    }
    return count;
}

int count_sentences(string text)
{
    int count = 0;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            count += 1;
        }
    }
    return count;
}

File: week-6/problem-sets/readability.py
def main():
    text = input("Text: ")
    L = float(count_letters(text) / count_words(text)) * 100.00
    S = float(count_sentences(text) / count_words(text)) * 100.00
    index = round((0.0588 * L) - (0.296 * S) - 15.8)
    if index >= 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


def count_letters(text):
    count = 0
    for char in text:
        if char.isalpha():
            count += 1
    return count


def count_words(text):
    # Accounts for the last word (i.e. without count=1, 'how are you' counts as two whereas 'how are you ' counts as three.)
    count = 1
    i = 0
    n = len(text)
    while i < n:
        if text[i] == ' ' or text[i] == '\n':
            while i + 1 < n and (text[i+1] == ' ' or text[i+1] == '\n'):
                i += 1
            count += 1
        i += 1
    return count


def count_sentences(text):
    count = 0
    i = 0
    n = len(text)
    while i < n:
        if text[i] == '.' or text[i] == '!' or text[i] == '?':
            count += 1
        i += 1
    return count


if __name__ == "__main__":
    main()
